tests/testmplex.nim

```diff
@@ -376,6 +378,77 @@ suite "Mplex":
   #   expect LPStreamEOFError:
   #     waitFor(testClosedForRead())
 
+  test "jitter - channel should be able to handle erratic read/writes":
+    proc test(): Future[bool] {.async.} =
+      let ma: MultiAddress = Multiaddress.init("/ip4/0.0.0.0/tcp/0")
+
+      var complete = newFuture[void]()
+      const MsgSize = 512 * 10
+      proc connHandler(conn: Connection) {.async, gcsafe.} =
+        proc handleMplexListen(stream: Connection) {.async, gcsafe.} =
+          let msg = await stream.readLp()
+          check msg.len == MsgSize
+          await stream.close()
+          complete.complete()
+
+        let mplexListen = newMplex(conn)
+        mplexListen.streamHandler = handleMplexListen
+        discard mplexListen.handle()
+
+      let transport1: TcpTransport = newTransport(TcpTransport)
+      discard await transport1.listen(ma, connHandler)
+
+      defer:
+        await transport1.close()
+
+      let transport2: TcpTransport = newTransport(TcpTransport)
+      let conn = await transport2.dial(transport1.ma)
+
+      let mplexDial = newMplex(conn)
+      let stream = await mplexDial.newStream()
+      var bigseq = newSeqOfCap[uint8](MaxMsgSize + 1)
+      for _ in 0..<MsgSize: # write one less than max size
+        bigseq.add(uint8(rand(uint('A')..uint('z'))))
+
+      ## create lenght prefixed libp2p frame
+      var buf = initVBuffer()
+      buf.writeSeq(bigseq)
+      buf.finish()
+
+      ## create mplex header
+      var mplexBuf = initVBuffer()
+      mplexBuf.writePBVarint((1.uint shl 3) or ord(MessageType.MsgOut).uint)
+      mplexBuf.writePBVarint(buf.buffer.len.uint) # size should be always sent
+
+      await conn.write(mplexBuf.buffer) # write out mplex header
+      proc writer() {.async.} =
+        var sent = 0
+        randomize()
+        let total = buf.buffer.len
+        const max = 500
+        const min = 100
```

Comment:
why not 1?

Reply:
too little and it isn't representative of real conditions.

Reply:
clarification, this is sending (randomized) variable size packets.

Reply:
of course it's representative: if you get fragmentation, the remainder number of bytes in the next tcp packet may very well be 1.

it would almost make sense to add a 1-byte trickle test for this reason. Consider varint for example, you will never stress the partial decoding of a 2-byte varint this way.

Reply:
ok, we can add a separate test for that, but I think this one stands on its own as well.

Reply:
added the 1 byte test